// node-simconnect/dist/enums/FacilityDataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FacilityDataType = void 0;
var FacilityDataType;
(function (FacilityDataType) {
    FacilityDataType[FacilityDataType["AIRPORT"] = 0] = "AIRPORT";
    FacilityDataType[FacilityDataType["RUNWAY"] = 1] = "RUNWAY";
    FacilityDataType[FacilityDataType["START"] = 2] = "START";
    FacilityDataType[FacilityDataType["FREQUENCY"] = 3] = "FREQUENCY";
    FacilityDataType[FacilityDataType["HELIPAD"] = 4] = "HELIPAD";
    FacilityDataType[FacilityDataType["APPROACH"] = 5] = "APPROACH";
    FacilityDataType[FacilityDataType["APPROACH_TRANSITION"] = 6] = "APPROACH_TRANSITION";
    FacilityDataType[FacilityDataType["APPROACH_LEG"] = 7] = "APPROACH_LEG";
    FacilityDataType[FacilityDataType["FINAL_APPROACH_LEG"] = 8] = "FINAL_APPROACH_LEG";
    FacilityDataType[FacilityDataType["MISSED_APPROACH_LEG"] = 9] = "MISSED_APPROACH_LEG";
    FacilityDataType[FacilityDataType["DEPARTURE"] = 10] = "DEPARTURE";
    FacilityDataType[FacilityDataType["ARRIVAL"] = 11] = "ARRIVAL";
    FacilityDataType[FacilityDataType["RUNWAY_TRANSITION"] = 12] = "RUNWAY_TRANSITION";
    FacilityDataType[FacilityDataType["ENROUTE_TRANSITION"] = 13] = "ENROUTE_TRANSITION";
    FacilityDataType[FacilityDataType["TAXI_POINT"] = 14] = "TAXI_POINT";
    FacilityDataType[FacilityDataType["TAXI_PARKING"] = 15] = "TAXI_PARKING";
    FacilityDataType[FacilityDataType["TAXI_PATH"] = 16] = "TAXI_PATH";
    FacilityDataType[FacilityDataType["TAXI_NAME"] = 17] = "TAXI_NAME";
    FacilityDataType[FacilityDataType["JETWAY"] = 18] = "JETWAY";
    FacilityDataType[FacilityDataType["VOR"] = 19] = "VOR";
    FacilityDataType[FacilityDataType["NDB"] = 20] = "NDB";
    FacilityDataType[FacilityDataType["WAYPOINT"] = 21] = "WAYPOINT";
    FacilityDataType[FacilityDataType["ROUTE"] = 22] = "ROUTE";
    FacilityDataType[FacilityDataType["PAVEMENT"] = 23] = "PAVEMENT";
    FacilityDataType[FacilityDataType["APPROACHLIGHTS"] = 24] = "APPROACHLIGHTS";
    FacilityDataType[FacilityDataType["VASI"] = 25] = "VASI";
})(FacilityDataType = exports.FacilityDataType || (exports.FacilityDataType = {}));
module.exports = {
    FacilityDataType,
};
