// node-simconnect/dist/connectionParameters.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.autodetectServerAddress = void 0;
const Path = require("path");
const os = require("os");
const debug_1 = require("debug");
const ini_1 = require("./utils/ini");
const registry_1 = require("./utils/registry");
const network_1 = require("./utils/network");
const logger = (0, debug_1.default)('node-simconnect');
function findSimConnectPortIPv4() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const port = yield (0, registry_1.readRegistryValue)('HKCU\\Software\\Microsoft\\Microsoft Games\\Flight Simulator', 'SimConnect_Port_IPv4');
            if (!port) {
                throw new Error('Could not find SimConnect_Port_IPv4 in the Windows registry');
            }
            return parseInt(port, 10);
        }
        catch (_a) {
            return 2048;
        }
    });
}
function readNetworkConfigFromSimConnectCfg(folderPath, index) {
    return __awaiter(this, void 0, void 0, function* () {
        const filePath = Path.join(folderPath, 'SimConnect.cfg');
        let fullCfg;
        try {
            // SimConnect.cfg uses the INI fileformat
            fullCfg = yield (0, ini_1.readIniFile)(filePath);
        }
        catch (e) {
            logger('Could not read SimConnect.cfg due to to the following error:', e);
            return undefined;
        }
        if (fullCfg.SimConnect === undefined) {
            throw new Error(`Invalid SimConnect.cfg file: ${filePath}`);
        }
        const indexStr = index !== undefined ? index.toString(10) : '0';
        const cfg = fullCfg.SimConnect[indexStr] ? fullCfg.SimConnect[indexStr] : fullCfg.SimConnect;
        if (cfg.Protocol === undefined || cfg.Address === undefined || cfg.Port === undefined) {
            throw new Error(`The loaded SimConnect.cfg (${filePath}) is missing required parameters.`);
        }
        else if (cfg.Protocol && cfg.Protocol.toUpperCase() !== 'IPV4') {
            throw new Error('Only the Ipv4 protocol is supported at the moment');
        }
        return {
            type: 'ipv4',
            host: cfg.Address,
            port: parseInt(cfg.Port, 10),
        };
    });
}
function autodetectServerAddress(cfgIndex) {
    return __awaiter(this, void 0, void 0, function* () {
        // Check for SimConnect.cfg in current dir
        const localConfig = yield readNetworkConfigFromSimConnectCfg(process.cwd(), cfgIndex);
        if (localConfig)
            return localConfig;
        const homeConfig = yield readNetworkConfigFromSimConnectCfg(os.homedir(), cfgIndex);
        if (homeConfig)
            return homeConfig;
        if (cfgIndex !== undefined) {
            throw new Error(`No SimConnect.cfg file containing the given config index ${cfgIndex} was found`);
        }
        // Check if named pipe exist
        const PIPE = '\\\\.\\pipe\\Microsoft Flight Simulator\\SimConnect';
        const msfsSimconnectPipeOk = yield (0, network_1.checkIfNamedPipeExist)(PIPE);
        if (msfsSimconnectPipeOk) {
            return { type: 'pipe', address: PIPE };
        }
        // Read port number from Windows registry
        const ipv4port = yield findSimConnectPortIPv4();
        return { type: 'ipv4', host: 'localhost', port: ipv4port };
    });
}
exports.autodetectServerAddress = autodetectServerAddress;
