// node-simconnect/dist/SimConnectPacketBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimConnectPacketBuilder = void 0;
const RawBuffer_1 = require("./RawBuffer");
class SimConnectPacketBuilder {
    constructor(packetTypeId, protocol, packetDataBuffer) {
        packetDataBuffer === null || packetDataBuffer === void 0 ? void 0 : packetDataBuffer.clear(); // Prepare for new message
        /**
         * Packet header content (16 bytes):
         *    0-3    packet size (set later)
         *    4-7    protocol
         *    8-11   packet type / SimConnect function
         *    12-15  packet id (set later)
         */
        this.packetContent = packetDataBuffer || new RawBuffer_1.RawBuffer(256);
        this.packetContent.writeInt32(protocol, 4);
        this.packetContent.writeInt32(0xf0000000 | packetTypeId, 8);
        this.packetContent.setOffset(16);
    }
    putFloat32(value) {
        this.packetContent.writeFloat32(value);
        return this;
    }
    putFloat64(value) {
        this.packetContent.writeFloat64(value);
        return this;
    }
    putString(value, fixedLength) {
        this.packetContent.writeString(value, fixedLength);
        return this;
    }
    putString256(value) {
        this.packetContent.writeString256(value);
        return this;
    }
    putInt16(value, offset) {
        this.packetContent.writeInt16(value, offset);
        return this;
    }
    putInt32(value, offset) {
        this.packetContent.writeInt32(value, offset);
        return this;
    }
    putUint32(value, offset) {
        this.packetContent.writeUint32(value, offset);
        return this;
    }
    putUint64(value, offset) {
        this.packetContent.writeUint64(value, offset);
        return this;
    }
    putByte(value) {
        this.packetContent.writeByte(value);
        return this;
    }
    putBytes(value) {
        this.packetContent.write(value);
        return this;
    }
    getRawBuffer() {
        return this.packetContent;
    }
    /**
     * Finalize
     * @param sendId - of packet (can be used to identify packet when exception event occurs)
     */
    build(sendId) {
        const packetSize = this.packetContent.getOffset();
        // Finish packet header
        this.packetContent.writeInt32(packetSize, 0);
        this.packetContent.writeInt32(sendId, 12);
        return this.packetContent.getBuffer();
    }
}
exports.SimConnectPacketBuilder = SimConnectPacketBuilder;
